Add GNOME3 ISO, Don't run graphical installer ISOs as root

worldofpeace opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#sec-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change
I think it's a great idea to provide another graphical iso.
I've seen people turn away from nixos because they somehow thought the only DE
choice was plasma5.

This also renames the regular `iso_graphical` image to `iso_graphical_plasma5`.
We already definitely have to organize updating
* https://nixos.org/nixos/download.html

to accommodate this.

Some of you may be interested in the ISO size and it's
* 1.5 GB

---

Probably most important commit here is https://github.com/NixOS/nixpkgs/commit/d3c94894de1bf7a0b219725a337f3423a1f1c661.

The graphical installer will no longer use root, but a less privileged `live` user.
I'm going to quote the commit message here
```
There's many reason why it is and is going to
continue to be difficult to do this:

1. All display-managers (excluding slim) default PAM rules
   disallow root auto login.

2. We can't use wayland

3. We have to use system-wide pulseaudio

4. It could break applications in the session.
   This happened to dolphin in plasma5
   in the past.

```

Because of `1` and `2` it completely prevented me from doing this.
This change was already seen in https://github.com/NixOS/nixpkgs/pull/42610 and my opinion was received as reactive. 
I hope this better supports my claim, this is going to be more work for us in the long run to maintain.

Now you could say the ISO should just use slim, but this project is abandoned and we really shouldn't be using it. Not to mention it wouldn't integrate at all with gnome.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

I've ran the gnome and plasma5 iso's in qemu and they seemed to work fine.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

worldofpeace committed:
nixos/release: add GNOME3 ISO

Note we're not using wayland default in the graphical media because it
could cause headaches for Nvidia users. But the session is still available.

worldofpeace committed:
nixos/release: rename iso_graphical to iso_graphical_plasma5

And all the other things and paths to match it

worldofpeace committed:
installer: Don't run graphical installer ISOs as root

There's many reason why it is and is going to
continue to be difficult to do this:

1. All display-managers (excluding slim) default PAM rules
   disallow root auto login.

2. We can't use wayland

3. We have to use system-wide pulseaudio

4. It could break applications in the session.
   This happened to dolphin in plasma5
   in the past.

This is a growing technical debt, let's just use
passwordless sudo.

lheckemann commented on nixos/doc/manual/installation/installing.xml:
```diff
@@ -29,13 +29,14 @@
   <para>
    If you downloaded the graphical ISO image, you can run <command>systemctl
-   start display-manager</command> to start KDE. If you want to continue on the
+   start display-manager</command> to start the DE. If you want to continue on the
```
I think we should write this out instead of using a relatively rare abbreviation — maybe "the graphical environment".

lheckemann commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5.nix:
```diff
@@ -1,5 +1,5 @@
 # This module defines a NixOS installation CD that contains X11 and
-# Plasma5.
+# KDE 5.
```
"Plasma 5" or "KDE Plasma 5" is the correct name, there is no such thing as KDE 5.

worldofpeace replied:
Yeah that got there by accident

lheckemann commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-plasma5.nix:
```diff
@@ -30,15 +30,27 @@ with lib;
       Type=Application
       Name=NixOS Manual
-      Exec=firefox ${config.system.build.manual.manualHTMLIndex}
+      Exec=firefox ${config.system.build.manual.manual}/share/doc/nixos/index.html
```
Did the `manualHTMLIndex` variant not work?

Thought on the side: maybe this should be extracted into the nixos manual module…

lheckemann reviewed (approved):
Overall I like it!

One concern I have is that this would add another big file to the release, and @edolstra may not like that.

edolstra commented on nixos/release.nix:
```diff
@@ -149,9 +149,15 @@ in rec {
-  iso_graphical = forMatchingSystems [ "x86_64-linux" ] (system: makeIso {
-    module = ./modules/installer/cd-dvd/installation-cd-graphical-kde.nix;
-    type = "graphical";
+  iso_graphical_plasma5 = forMatchingSystems [ "x86_64-linux" ] (system: makeIso {
```
Note that renaming `iso_graphical` will break the NixOS channel generation script.

edolstra commented:
I'm not too keen on providing multiple graphical ISOs. Apart from the overhead this adds to the release process (in terms of storage costs and upload time), I'd rather have one well-tested ISO than multiple poorly tested ones. It also adds friction for new users since they have to decide which ISO to download, and it's not obvious which one is better.

edolstra commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-base.nix:
```diff
@@ -8,6 +8,30 @@ with lib;
 
+  users.extraUsers.live = {
+    isNormalUser = true;
+    uid = 1000;
```
We probably shouldn't use UIDs in the `>1000` range since that's reserved for users. Maybe just leave out the UID?

edolstra commented on nixos/modules/installer/cd-dvd/installation-cd-graphical-base.nix:
```diff
@@ -8,6 +8,30 @@ with lib;
+  users.extraUsers.live = {
+    isNormalUser = true;
+    uid = 1000;
+    extraGroups = [ "wheel" "networkmanager" "video" ];
```
Are `networkmanager` and `video` needed? NetworkManager should work for locally logged-in users without it, and likewise logged-in users get access for /dev/video* via an ACL.

worldofpeace replied:
I'm not sure. I think I'd have an issue on my own installation if I didn't use this. I've received guidance that this should be done, but I've yet to verify this.

I will find out :+1:

worldofpeace replied:
I tried this in a vm and it caused it to require authentication to initiate a connection with networkmanager. Not sure about `video` though but perhaps it's fine to leave it.

edolstra commented:
We could merge this without the changes to `nixos/release.nix` for now.

worldofpeace commented:
> @edolstra 
> I'm not too keen on providing multiple graphical ISOs. Apart from the overhead this adds to the release process (in terms of storage costs and upload time), I'd rather have one well-tested ISO than multiple poorly tested ones. It also adds friction for new users since they have to decide which ISO to download, and it's not obvious which one is better.

Are we saying that manually testing these is the way to go?
Because if we had good automated tested for these it shouldn't be  a problem how many ISO's we distribute, aside from the technical overhead. Additionally, I've put a lot of work into GNOME3 in NixOS, I'd like my work to be visible to users trying it.  It's a good way to attract desktop users in general.

> @lheckemann 
> One concern I have is that this would add another big file to the release

Indeed it is a big file, but it's still smaller than other GNOME3 based iso's in the wild.

Examples
* [Ubuntu 19.04](https://ubuntu.com/download/desktop/thank-you/?version=19.04&architecture=amd64) (2.1 GB)
* [Fedora Workstation 30](https://getfedora.org/en/workstation/download/) (1.9)
* [Manjaro](https://osdn.net/dl/manjaro/manjaro-gnome-18.0.4-stable-x86_64.iso) (2.0 GB)
* [Pop!_OS]( https://pop-iso.sfo2.cdn.digitaloceanspaces.com/19.04/amd64/intel/10/pop-os_19.04_amd64_intel_10.iso) (2.1 GB)

Though [Opensuse Leap](https://software.opensuse.org/distributions/leap) has somehow got it to 921 MB.

So I'd say, comparatively, it's in line with being one of the smaller ones.

> @edolstra 
> We could merge this without the changes to nixos/release.nix for now.

I can see how we wouldn't want to block this on that, I respect your guidance if you think the technical overhead should cause us to consider carefully adding this to the release.

samueldr reviewed (changes requested):
We need **all** installer images to run under the same user. The minimal one included.

Otherwise, this may cause issues in writing documentation, and supporting users. Assumptions will be made "they are running ____" but instead they are running ____, thus the assumption doesn't hold and they need to either run through sudo or not.

Though, I'm not saying that graphical images must be run as root, only that all installers should use the same user, and relevant installation notes fixed, if needs be.

worldofpeace commented:
@samueldr Ah I overlooked that, I will move the user to the base profile etc.

Note I think the ova image uses a `demo` user already, but that was inconsistent before this.

worldofpeace committed:
fixup! installer: Don't run graphical installer ISOs as root

* Move the config to installation-device.nix
* correct mingetty helpline

polkit config is still in graphical-base because
that is where polkit is useful.

worldofpeace committed:
fixup! fixup! installer: Don't run graphical installer ISOs as root

* Don't use old users.extraUsers option.
* Don't set uid.

worldofpeace committed:
fixup! fixup! fixup! installer: Don't run graphical installer ISOs as root

Fixup the release note.

worldofpeace committed:
fixup! fixup! fixup! fixup! installer: Don't run graphical installer ISOs as root

worldofpeace commented:
So I've decided with some guidance from @grahamc that it would help me to split this into smaller PR's.

First one will be `Don't run installer as root`.
The other stuff will come after since this needs that change first.
